### ISSUE TITLE
Ansible facts override special variables

### ISSUE THREAD

**Issue by NomAnor** [ansible/ansible#16935]
##### ISSUE TYPE
- Bug Report
##### ANSIBLE VERSION

ansible 2.1.1.0
  configured module search path = Default w/o overrides
##### OS / ENVIRONMENT

Linux Mint 17
##### SUMMARY

Ansible facts generated by the setup module will override special variables like _ansible_user_.
##### STEPS TO REPRODUCE

I have a server wich acts as a router and I renamed the network interfaces so it's easier to see which interface belongs to which network. So instead of  _ethX_ one of my interfaces is called _user_. 
##### EXPECTED RESULTS

The setup module should not create "harmful" fact names.
##### ACTUAL RESULTS

After ansible gathered the facts the _ansible_user_ variable gets overridden with the information of that interface and later tasks can't connect with ssh because of a wrong username.

It would be better to use a dictionary like _ansible_devices_ e.g _ansible_interfaces_.

**Comment by jctanner** [comment]
@NomAnor can you provide an example of the facts that come back from the host?

**Comment by NomAnor** [comment]
This is the output of  `ip addr`:

```
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default 
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: user: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:3b:41:db brd ff:ff:ff:ff:ff:ff
    inet 192.168.100.2/24 brd 192.168.100.255 scope global user
       valid_lft forever preferred_lft forever
    inet6 fe80::5054:ff:fe3b:41db/64 scope link 
       valid_lft forever preferred_lft forever
```

These are the facts gathered with `ansible -m setup -i 192.168.100.2, -u root -k all`:

```
192.168.100.2 | SUCCESS => {
    "ansible_facts": {
        "ansible_all_ipv4_addresses": [
            "192.168.100.2"
        ], 
        "ansible_all_ipv6_addresses": [
            "fe80::5054:ff:fe3b:41db"
        ], 
        "ansible_architecture": "x86_64", 
        "ansible_bios_date": "01/01/2011", 
        "ansible_bios_version": "Bochs", 
        "ansible_cmdline": {
            "BOOT_IMAGE": "/boot/vmlinuz-3.16.0-4-amd64", 
            "quiet": true, 
            "ro": true, 
            "root": "UUID=edcd9fd2-1c10-490f-84a3-ada93ddb8d4f"
        }, 
        "ansible_date_time": {
            "date": "2016-08-10", 
            "day": "10", 
            "epoch": "1470819167", 
            "hour": "08", 
            "iso8601": "2016-08-10T08:52:47Z", 
            "iso8601_basic": "20160810T085247963687", 
            "iso8601_basic_short": "20160810T085247", 
            "iso8601_micro": "2016-08-10T08:52:47.963804Z", 
            "minute": "52", 
            "month": "08", 
            "second": "47", 
            "time": "08:52:47", 
            "tz": "UTC", 
            "tz_offset": "+0000", 
            "weekday": "Mittwoch", 
            "weekday_number": "3", 
            "weeknumber": "32", 
            "year": "2016"
        }, 
        "ansible_default_ipv4": {
            "address": "192.168.100.2", 
            "alias": "user", 
            "broadcast": "192.168.100.255", 
            "gateway": "192.168.100.1", 
            "interface": "user", 
            "macaddress": "52:54:00:3b:41:db", 
            "mtu": 1500, 
            "netmask": "255.255.255.0", 
            "network": "192.168.100.0", 
            "type": "ether"
        }, 
        "ansible_default_ipv6": {}, 
        "ansible_devices": {
            "sda": {
                "holders": [], 
                "host": "SCSI storage controller: Red Hat, Inc Virtio SCSI", 
                "model": "QEMU HARDDISK", 
                "partitions": {
                    "sda1": {
                        "sectors": "1998848", 
                        "sectorsize": 512, 
                        "size": "976.00 MB", 
                        "start": "2048"
                    }, 
                    "sda2": {
                        "sectors": "39940096", 
                        "sectorsize": 512, 
                        "size": "19.04 GB", 
                        "start": "2000896"
                    }
                }, 
                "removable": "0", 
                "rotational": "1", 
                "sas_address": null, 
                "sas_device_handle": null, 
                "scheduler_mode": "cfq", 
                "sectors": "41943040", 
                "sectorsize": "512", 
                "size": "20.00 GB", 
                "support_discard": "4096", 
                "vendor": "QEMU"
            }
        }, 
        "ansible_distribution": "Debian", 
        "ansible_distribution_major_version": "8", 
        "ansible_distribution_release": "jessie", 
        "ansible_distribution_version": "8.5", 
        "ansible_dns": {
            "domain": "internal", 
            "nameservers": [
                "127.0.0.1"
            ]
        }, 
        "ansible_domain": "internal", 
        "ansible_env": {
            "HOME": "/root", 
            "LANG": "de_DE.UTF-8", 
            "LC_ALL": "de_DE.UTF-8", 
            "LC_CTYPE": "de_DE.UTF-8", 
            "LC_MESSAGES": "de_DE.UTF-8", 
            "LOGNAME": "root", 
            "MAIL": "/var/mail/root", 
            "PATH": "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin", 
            "PWD": "/root", 
            "SHELL": "/bin/bash", 
            "SHLVL": "1", 
            "SSH_CLIENT": "192.168.100.1 48381 22", 
            "SSH_CONNECTION": "192.168.100.1 48381 192.168.100.2 22", 
            "SSH_TTY": "/dev/pts/1", 
            "TERM": "screen", 
            "USER": "root", 
            "_": "/bin/sh"
        }, 
        "ansible_fips": false, 
        "ansible_form_factor": "Other", 
        "ansible_fqdn": "test.internal", 
        "ansible_gather_subset": [
            "hardware", 
            "network", 
            "virtual"
        ], 
        "ansible_hostname": "test", 
        "ansible_interfaces": [
            "lo", 
            "user"
        ], 
        "ansible_kernel": "3.16.0-4-amd64", 
        "ansible_lo": {
            "active": true, 
            "device": "lo", 
            "ipv4": {
                "address": "127.0.0.1", 
                "broadcast": "host", 
                "netmask": "255.0.0.0", 
                "network": "127.0.0.0"
            }, 
            "ipv6": [
                {
                    "address": "::1", 
                    "prefix": "128", 
                    "scope": "host"
                }
            ], 
            "mtu": 65536, 
            "promisc": false, 
            "type": "loopback"
        }, 
        "ansible_machine": "x86_64", 
        "ansible_machine_id": "f412d0fba4ce4931b5106b7a72f7c666", 
        "ansible_memfree_mb": 790, 
        "ansible_memory_mb": {
            "nocache": {
                "free": 935, 
                "used": 65
            }, 
            "real": {
                "free": 790, 
                "total": 1000, 
                "used": 210
            }, 
            "swap": {
                "cached": 0, 
                "free": 975, 
                "total": 975, 
                "used": 0
            }
        }, 
        "ansible_memtotal_mb": 1000, 
        "ansible_mounts": [
            {
                "device": "/dev/sda2", 
                "fstype": "ext4", 
                "mount": "/", 
                "options": "rw,relatime,errors=remount-ro,data=ordered", 
                "size_available": 17529397248, 
                "size_total": 19994112000, 
                "uuid": "edcd9fd2-1c10-490f-84a3-ada93ddb8d4f"
            }
        ], 
        "ansible_nodename": "test", 
        "ansible_os_family": "Debian", 
        "ansible_pkg_mgr": "apt", 
        "ansible_processor": [
            "GenuineIntel", 
            "QEMU Virtual CPU version 2.0.0", 
            "GenuineIntel", 
            "QEMU Virtual CPU version 2.0.0"
        ], 
        "ansible_processor_cores": 1, 
        "ansible_processor_count": 2, 
        "ansible_processor_threads_per_core": 1, 
        "ansible_processor_vcpus": 2, 
        "ansible_product_name": "Standard PC (i440FX + PIIX, 1996)", 
        "ansible_product_serial": "NA", 
        "ansible_product_uuid": "18FFA67D-B671-3B21-7652-E057E62C8E1C", 
        "ansible_product_version": "pc-i440fx-trusty", 
        "ansible_python": {
            "executable": "/usr/bin/python", 
            "has_sslcontext": true, 
            "type": "CPython", 
            "version": {
                "major": 2, 
                "micro": 9, 
                "minor": 7, 
                "releaselevel": "final", 
                "serial": 0
            }, 
            "version_info": [
                2, 
                7, 
                9, 
                "final", 
                0
            ]
        }, 
        "ansible_python_version": "2.7.9", 
        "ansible_selinux": false, 
        "ansible_service_mgr": "systemd", 
        "ansible_ssh_host_key_dsa_public": "AAAAB3NzaC1kc3MAAACBAOASh5ilpumfBvhEHP4Qm9bnItq2qvUIjUbxOtacwliHELbF+nOSvkXstQZKiObBFEP+XkOSdmS3J52TD2uopTUeZ9FqsZYf+8uqE8/ay0khGzJRYfhj3puU06SDYdSe3bIhTnABrc/GGGPFvvP1gwoK/8jwuMPWU4H8yiMawNcvAAAAFQCLFyUS1fCkQ2It62B4Rx+FSt2Y7wAAAIEAitkKeAAYckyMK7ifNEcyyV8hMcuXu2lwKbJDG73yeJM4phi1TSWrNkAjtn+n+w78DOWh0slt7aHxzv8EDuux8Kez3Lj0nvzDZP+xQYN5LOWAr3PIksJKshz0RGCC3gVCLFas4gO63QrRJooKuUjnW6JA/QUOkUyn9aUG+C1edEcAAACAILi5eKhfd8x8kpxa7LtPOJWZpmgu5/WsnKiTUiiBBjvQ2EvKgu5Ymis1VQi5mMvUjm3cZepQ5NxeWqTAtjXjhXtjswvuFhJ9mSAVcLeK9sxobQG7mKXK9JS2llByI8s8/0huHJ3GdGrD7C5s1pKRXj3a3+f4S3W0qPvjLwFtxoc=", 
        "ansible_ssh_host_key_ecdsa_public": "AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBLyWInLuwe1D/Bx104T/mm0HznShAA+3A0/668Kt6BERmZmWPRlFPpsk0ZbjcImPTmKbA6sGnZwZ7yKi0Ero4Xs=", 
        "ansible_ssh_host_key_ed25519_public": "AAAAC3NzaC1lZDI1NTE5AAAAIJwWJDy8N2n9cOVEflx7S0lwMjpFHq2E3CGj530PCo0m", 
        "ansible_ssh_host_key_rsa_public": "AAAAB3NzaC1yc2EAAAADAQABAAABAQCW56FEDW7OPdtt7U7p7GQKPsgPLXe0Dx6kYeG0/9wsQVW2b+LGMa1CnoSdHXgnsxSUcyOIveFqIVG0c5iZYcnEAgIN3MsOkqp0CcT8qaSGdE2RY1c8+RBZDTCFRj+s6L+t1NNjQbHLH1QjtvwoHodPf+7S+85wKG1PZceSuTx/ra75wD9QBoIY9gaeJoqwtCA9WURGTcKDSzcdfMAPH1lr0Pk5SbEY/E7OSLK3J4xniPGtdAWr1hruY9D9/IIqZ+6EcTCMXZaTbLlPfBbITcrbtjQL8OSRlwgvKxh7E2dF8Clyhmvqtbl4jdbQqksomRmaJ58o1Y5Fb3o6MuBosC1r", 
        "ansible_swapfree_mb": 975, 
        "ansible_swaptotal_mb": 975, 
        "ansible_system": "Linux", 
        "ansible_system_capabilities": [
            "cap_chown", 
            "cap_dac_override", 
            "cap_dac_read_search", 
            "cap_fowner", 
            "cap_fsetid", 
            "cap_kill", 
            "cap_setgid", 
            "cap_setuid", 
            "cap_setpcap", 
            "cap_linux_immutable", 
            "cap_net_bind_service", 
            "cap_net_broadcast", 
            "cap_net_admin", 
            "cap_net_raw", 
            "cap_ipc_lock", 
            "cap_ipc_owner", 
            "cap_sys_module", 
            "cap_sys_rawio", 
            "cap_sys_chroot", 
            "cap_sys_ptrace", 
            "cap_sys_pacct", 
            "cap_sys_admin", 
            "cap_sys_boot", 
            "cap_sys_nice", 
            "cap_sys_resource", 
            "cap_sys_time", 
            "cap_sys_tty_config", 
            "cap_mknod", 
            "cap_lease", 
            "cap_audit_write", 
            "cap_audit_control", 
            "cap_setfcap", 
            "cap_mac_override", 
            "cap_mac_admin", 
            "cap_syslog", 
            "cap_wake_alarm", 
            "cap_block_suspend", 
            "37+ep"
        ], 
        "ansible_system_capabilities_enforced": "True", 
        "ansible_system_vendor": "QEMU", 
        "ansible_uptime_seconds": 174, 
        "ansible_user": {
            "active": true, 
            "device": "user", 
            "ipv4": {
                "address": "192.168.100.2", 
                "broadcast": "192.168.100.255", 
                "netmask": "255.255.255.0", 
                "network": "192.168.100.0"
            }, 
            "ipv6": [
                {
                    "address": "fe80::5054:ff:fe3b:41db", 
                    "prefix": "64", 
                    "scope": "link"
                }
            ], 
            "macaddress": "52:54:00:3b:41:db", 
            "module": "e1000", 
            "mtu": 1500, 
            "pciid": "0000:00:03.0", 
            "promisc": false, 
            "type": "ether"
        }, 
        "ansible_user_dir": "/root", 
        "ansible_user_gecos": "root", 
        "ansible_user_gid": 0, 
        "ansible_user_id": "root", 
        "ansible_user_shell": "/bin/bash", 
        "ansible_user_uid": 0, 
        "ansible_userspace_architecture": "x86_64", 
        "ansible_userspace_bits": "64", 
        "ansible_virtualization_role": "guest", 
        "ansible_virtualization_type": "kvm", 
        "module_setup": true
    }, 
    "changed": false
}
```

and this is the error message for a playbook with one tasks:

```
PLAYBOOK: playbook.yml *********************************************************
1 plays in playbook.yml

PLAY [Test] ********************************************************************

TASK [setup] *******************************************************************
<192.168.100.2> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.100.2> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.100.2 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1470819206.23-68065336390587 `" && echo ansible-tmp-1470819206.23-68065336390587="` echo $HOME/.ansible/tmp/ansible-tmp-1470819206.23-68065336390587 `" ) && sleep 0'"'"''
<192.168.100.2> PUT /tmp/tmpZEhuAp TO /root/.ansible/tmp/ansible-tmp-1470819206.23-68065336390587/setup
<192.168.100.2> SSH: EXEC sshpass -d12 sftp -o BatchMode=no -b - -C -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r '[192.168.100.2]'
<192.168.100.2> ESTABLISH SSH CONNECTION FOR USER: root
<192.168.100.2> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o User=root -o ConnectTimeout=10 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r -tt 192.168.100.2 '/bin/sh -c '"'"'LANG=de_DE.UTF-8 LC_ALL=de_DE.UTF-8 LC_MESSAGES=de_DE.UTF-8 /usr/bin/python /root/.ansible/tmp/ansible-tmp-1470819206.23-68065336390587/setup; rm -rf "/root/.ansible/tmp/ansible-tmp-1470819206.23-68065336390587/" > /dev/null 2>&1 && sleep 0'"'"''
ok: [192.168.100.2]

TASK [setup] *******************************************************************
task path: /home/xxx/playbook.yml:7
<192.168.100.2> ESTABLISH SSH CONNECTION FOR USER: {u'macaddress': u'52:54:00:3b:41:db', u'pciid': u'0000:00:03.0', u'module': u'e1000', u'mtu': 1500, u'device': u'user', u'promisc': False, u'ipv4': {u'broadcast': u'192.168.100.255', u'netmask': u'255.255.255.0', u'network': u'192.168.100.0', u'address': u'192.168.100.2'}, u'ipv6': [{u'scope': u'link', u'prefix': u'64', u'address': u'fe80::5054:ff:fe3b:41db'}], u'active': True, u'type': u'ether'}
<192.168.100.2> SSH: EXEC sshpass -d12 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o 'User={u'"'"'macaddress'"'"': u'"'"'52:54:00:3b:41:db'"'"', u'"'"'pciid'"'"': u'"'"'0000:00:03.0'"'"', u'"'"'module'"'"': u'"'"'e1000'"'"', u'"'"'mtu'"'"': 1500, u'"'"'device'"'"': u'"'"'user'"'"', u'"'"'promisc'"'"': False, u'"'"'ipv4'"'"': {u'"'"'broadcast'"'"': u'"'"'192.168.100.255'"'"', u'"'"'netmask'"'"': u'"'"'255.255.255.0'"'"', u'"'"'network'"'"': u'"'"'192.168.100.0'"'"', u'"'"'address'"'"': u'"'"'192.168.100.2'"'"'}, u'"'"'ipv6'"'"': [{u'"'"'scope'"'"': u'"'"'link'"'"', u'"'"'prefix'"'"': u'"'"'64'"'"', u'"'"'address'"'"': u'"'"'fe80::5054:ff:fe3b:41db'"'"'}], u'"'"'active'"'"': True, u'"'"'type'"'"': u'"'"'ether'"'"'}' -o ConnectTimeout=10 -o ControlPath=/home/xxx/.ansible/cp/ansible-ssh-%h-%p-%r 192.168.100.2 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1470819206.57-262805914468828 `" && echo ansible-tmp-1470819206.57-262805914468828="` echo $HOME/.ansible/tmp/ansible-tmp-1470819206.57-262805914468828 `" ) && sleep 0'"'"''
fatal: [192.168.100.2]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh.", "unreachable": true}

PLAY RECAP *********************************************************************
192.168.100.2              : ok=1    changed=0    unreachable=1    failed=0   
```

As you see after gathering facts the `ansible_user` variable is overridden and no further ssh connection can be etablished.

**Comment by bcoca** [comment]
fixed via 236c923c25c053b0ff912f328e87da2dd26cbf69
